// @u4/opencv-build/dist/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ALL_OPENCV_MODULES = exports.args2Option = exports.genHelp = exports.ALLARGS = exports.OpenCVBuildEnv = exports.getLibsFactory = exports.OpenCVBuilder = void 0;
const OpenCVBuilder_js_1 = __importDefault(require("./OpenCVBuilder.js"));
var OpenCVBuilder_js_2 = require("./OpenCVBuilder.js");
Object.defineProperty(exports, "OpenCVBuilder", { enumerable: true, get: function () { return __importDefault(OpenCVBuilder_js_2).default; } });
var getLibsFactory_js_1 = require("./getLibsFactory.js");
Object.defineProperty(exports, "getLibsFactory", { enumerable: true, get: function () { return getLibsFactory_js_1.getLibsFactory; } });
var OpenCVBuildEnv_js_1 = require("./OpenCVBuildEnv.js");
Object.defineProperty(exports, "OpenCVBuildEnv", { enumerable: true, get: function () { return __importDefault(OpenCVBuildEnv_js_1).default; } });
var misc_js_1 = require("./misc.js");
Object.defineProperty(exports, "ALLARGS", { enumerable: true, get: function () { return misc_js_1.ALLARGS; } });
Object.defineProperty(exports, "genHelp", { enumerable: true, get: function () { return misc_js_1.genHelp; } });
Object.defineProperty(exports, "args2Option", { enumerable: true, get: function () { return misc_js_1.args2Option; } });
Object.defineProperty(exports, "ALL_OPENCV_MODULES", { enumerable: true, get: function () { return misc_js_1.ALL_OPENCV_MODULES; } });
exports.default = OpenCVBuilder_js_1.default;
